Input example units and conversions

truongdd03 opened the pull request:
# Description

This PR inputs the example units/conversions in the document. Besides that, it also contains some minor changes in unit/conversion models/tests.

Fixes automated-input-of-example-units-and-conversions

Note this brings in the development branch via a merge. This is why the node_modules must be updated. Developers need to add --build the first time they bring OED up to get node to update (docker compose up --build).

## Type of change

- [x] Note merging this changes the node modules
- [x] Note merging this changes the database configuration.
- [ ] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

## Limitations

jina2k committed:
removed stub types

jina2k committed:
babel v6 to v7

jina2k committed:
ts-loader and fixed the syntax errors that existed for a while

jina2k committed:
uninstall awesome-typescript-loader and some syntax spacing

jina2k committed:
removed most componentWill commands, warnings still appear

jina2k committed:
react-intl v2 to v5

jina2k committed:
aligning with dev

jina2k committed:
trying to remove conflicts before merge

jina2k committed:
merged packageupdate2

jina2k committed:
Merge branch 'development' into development

jina2k committed:
successful run with merge after reinstall

jina2k committed:
Merge branch 'development' of https://github.com/jina2k/OED into development

jina2k committed:
working

jina2k committed:
everything but react-intl v2 to v5

jina2k committed:
lint check and revert docker change

jina2k committed:
Merge branch 'development' into otherdev

jina2k committed:
react intl v2 to v5 and revert ts to 4.4.4

jina2k committed:
removed warnings

jina2k committed:
react-plotlyjs-ts somewhat deprecated, replaced with react/plotly.js

jina2k committed:
static size of 900 for height, width set to 100% for plotly, height is only changed for both line and bar

jina2k committed:
Merge remote-tracking branch 'origin' into otherdev

jina2k committed:
wip of eslint

jina2k committed:
changing tab rules and removing semicolon rule

jina2k committed:
added commonjs for now, fixed mostly tab spacing

jina2k committed:
Merge branch 'development' into otherdev

huss committed:
fixes issue with Obvius config files

If a second config file is uploaded with the same meter then it now
checks so it does not try to recreate.
This was causing error msgs during the test suite.
It is unknown if Obvius will do this (probably) and a TODO indicates we need
to decide if the meter should be updated.

ngan-cassie committed:
display date range for compare graph

huss committed:
shorten dates on graphics

The moment format was changed so it uses fewer characters.
This is a small change but will help make hover and x-axis labels be
shorter so they take up less space.

Note that compare is not done since its labels are being changed.
It is assume that work will do the same changes as this.

huss committed:
Merge pull request #708 from huss/graphDates

shorten dates on graphics

huss committed:
add code to enable React traces

huss committed:
Merge pull request #707 from huss/configTest

fixes issue with Obvius config files

ngan-cassie committed:
test start timestamp compare vs bar

huss committed:
Merge pull request #709 from huss/reduceDev

add code to enable React traces

huss committed:
update dates and translate

ngan-cassie committed:
Merge pull request #1 from huss/compareLabel

update dates and translate

ngan-cassie committed:
Merge branch 'OpenEnergyDashboard:development' into development

ngan-cassie committed:
Merge branch 'development' of https://github.com/ngan-cassie/OED into development

ngan-cassie committed:
fix styling

huss committed:
update compare labels

- Add start/end x-axis title for both last time period and
this time period
- Remove from to/from make shorter
- Use - not to since consistent with bar

huss committed:
fixes bug with compareTimeInterval

OED uses compare times that are to the whole hour since the data is
gotten by the hour/day for comparisons. The code did not initially set
the end time of the compareTimeInterval to the hour. This had two
issues. First, for each new request, the time shifted in the hour and
the Redux state was recalculated to be saved even though it did not
change. This was inefficient. Second, when groups were involved,
the state was to the exact time but the used interval was to the hour
so an error occurred when the value was not found in the state. This
only happened at times but caused compare graphics to not display.

huss committed:
trigger GitHub actions

huss committed:
try to trigger GitHub Actions

For reasons I'm not sure of, the checks did not fire on my last changes.
I tried an empty commit but that did not help. Thus, I'm doing a basic
formatting change to see if that will help. The check and test work fine
on my machine.

huss committed:
Merge pull request #710 from ngan-cassie/display-start-and-end-times-for-compare-graph

Display start and end times for compare graph

jina2k committed:
Revert "added commonjs for now, fixed mostly tab spacing"

This reverts commit 2ddb9499bcd05785fa07b5c8c63de6410963b793.

jina2k committed:
Revert "changing tab rules and removing semicolon rule"

This reverts commit 87903a1085dad5545973d0d49857b5af0c12fe7c.

jina2k committed:
Revert "wip of eslint"

This reverts commit b0c4995156361e443b200d242acf7c61aa2fa82e.

jina2k committed:
reverted eslint for now

jina2k committed:
json issues with undefined

jina2k committed:
npm audit for nanoid

dependabot[bot] committed:
Bump follow-redirects from 1.14.7 to 1.14.8

Bumps [follow-redirects](https://github.com/follow-redirects/follow-redirects) from 1.14.7 to 1.14.8.
- [Release notes](https://github.com/follow-redirects/follow-redirects/releases)
- [Commits](https://github.com/follow-redirects/follow-redirects/compare/v1.14.7...v1.14.8)

---
updated-dependencies:
- dependency-name: follow-redirects
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

huss committed:
Merge pull request #713 from OpenEnergyDashboard:dependabot/npm_and_yarn/follow-redirects-1.14.8

Bump follow-redirects from 1.14.7 to 1.14.8

jina2k committed:
fixed error when changing bar stacking twice

huss committed:
add MPL to data.js

This file is now JS so can add the header.
Remove exclusion from check since now fine.

huss committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into pr/jina2k/696

huss committed:
Merge pull request #696 from jina2k/otherdev

Updating packages #2

Joseph-M-Newman committed:
internationalize GPS error popup message #705

GPS error popup now displays in proper languages, English, Spanish, French

Joseph-M-Newman committed:
Fix#705

Removed Old code, and formated Data.js a Little better. Changed intlError to input.gps.range

huss committed:
fix alphabetical order

huss committed:
Merge pull request #714 from Joseph-M-Newman/development

internationalize GPS error popup message Fix

huss committed:
update node to latest LTS

huss committed:
Merge pull request #716 from huss/huss

update node to latest LTS

jina2k committed:
fixing previous bugs with getDerivedStateFromProps

jina2k committed:
small change same meaning

huss committed:
Merge pull request #718 from jina2k/eslintreactb

fixing previous pr bugs with getDerivedStateFromProps

jina2k committed:
react-redux v5 to v6 wip

jina2k committed:
Merge branch 'OpenEnergyDashboard:development' into eslintreactb

huss committed:
update install to avoid npm installs

huss committed:
Merge pull request #721 from huss/nodeInstall

update install to avoid npm installs

jina2k committed:
fixed package

jina2k committed:
Merge branch 'OpenEnergyDashboard:development' into eslintreactb

jina2k committed:
redux-devtools-extension update

jina2k committed:
react-redux v7 wip

jina2k committed:
fixed issue with types

jina2k committed:
maps now work with groups and meters

jina2k committed:
showing updates for editgroup

huss committed:
Merge pull request #722 from jina2k:eslintreactb

React-redux v5 to v7

huss committed:
update how dropdown menus work

Some Dropdown menu choices were using outline incorrectly.
This fixes but also changes so the value selected is shown
rather than telling the user to select a value.

huss committed:
fix linting issue

huss committed:
Merge pull request #725 from huss/outlineIssue

update how dropdown menus work

truongdd03 committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into automated_test

truongdd03 committed:
Insert standard units and conversions

truongdd03 committed:
Add special units

truongdd03 committed:
Fix small bugs in tests and move insert funtions to createDB

truongdd03 committed:
Add commands to insert special units asnd conversions

truongdd03 committed:
Fix lint errors

truongdd03 committed:
Remove some units and conversions

truongdd03 committed:
Add temperature units and conversions

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -302,6 +305,68 @@ function generateVariableAmplitudeTestingData() {
 	}
 }
 
+/**
```
You cannot execute the conversion insertion unless the units have been done. Thus, I was thinking a new function that combined them might be good. It would first call units and then call conversions. Then, the package.json would be modified to just have a script to call this new function. This would eliminate the ability to insert just one of these but I'm not sure that is of real interest (and that would still be in the code).

huss commented on src/scripts/installOED.sh:
```diff
@@ -117,6 +127,9 @@ while [ $create_error == 0 ]; do
 		# There was no createdb error so assume database ready for use so stop process
 		create_error=1
 		printf "%s\n" "  database creation had no errors so assume schema creation worked."
+		# Insert special units and conversions
```
I don't think we want this here. The special units are for developers only. Thus, they should be added manually as they do with all other test data. If we leave this then production sites would also get these units.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -302,6 +305,68 @@ function generateVariableAmplitudeTestingData() {
+	for (let i = 0; i < specialUnits.length; ++i) {
+		const unitData = specialUnits[i];
+		if (await Unit.getByName(unitData[0], conn) === null) {
+			await new Unit(undefined, unitData[0], unitData[1], Unit.unitRepresentType.UNUSED, 3600, 
```
I think the 4th parameter should be Unit.unitRepresentType.QUANTITY for each one of type METER. At this point the examples only do quantity type readings. It is okay to apply to all since it is not used if not a type meter.

I would just leave the 5th blank instead of 3600 since the code will default to 3600 and would still be okay if that is changed in the future.

huss commented on src/server/data/automatedTestingData.js:
```diff
@@ -302,6 +305,68 @@ function generateVariableAmplitudeTestingData() {
+		const unitData = specialUnits[i];
+		if (await Unit.getByName(unitData[0], conn) === null) {
+			await new Unit(undefined, unitData[0], unitData[1], Unit.unitRepresentType.UNUSED, 3600, 
+				unitData[2], null, unitData[3], unitData[4], unitData[5], null).insert(conn);
```
Should an empty note be null or "". I'm uncertain what other code will do if it is null. In any case, maybe it should be "special unit" so it is clear in the DB that this is the case.

huss commented on src/server/models/Conversion.js:
```diff
@@ -32,6 +33,31 @@ class Conversion {
+			['MJ', 'M3_gas', true, 2.6e-2, 0, 'MJ → M3_gas'],
+			['MJ', 'BTU', true, 947.8, 0, 'MJ → BTU'],
+			['kg', 'Metric_ton', true, 1e-3, 0, 'kg → Metric ton'],
+			['Fahrenheit', 'Celsius', true, 1.8, 32, 'Fahrenheit → Celsius']
```
I think the temperature conversion is backward. It should be:
['Celsius', 'Fahrenheit', true, 1.8, 32, 'Celsius → Fahrenheit']

truongdd03 replied:
I think the current one is correct since` Fahrenheit = 1.8 * Celsius + 32`. The order is similar to line 47, where `kg = 1e-3 * Metric_ton`.

huss replied:
Maybe I have this wrong but I still have a concern. I think the correct conversion for line 47 is
kg = 1000 * Metric_ton
because there are 1000 kg in a metric ton. Thus, the conversion table is specifying the other way around of
Metric_ton = 1e-3 * kg
If that is true then the temperature conversion is backward.

truongdd03 replied:
My bad. You are correct. The conversion should be reversed. Thank you for your explanation.

huss commented on src/server/models/Unit.js:
```diff
@@ -80,6 +80,33 @@ class Unit {
+		for (let i = 0; i < standardUnits.length; ++i) {
+			const unitData = standardUnits[i];
+			if (await Unit.getByName(unitData[0], conn) === null) {
+				await new Unit(undefined, unitData[0], unitData[1], Unit.unitRepresentType.UNUSED, 3600, 
```
See comments in automatedTestingData.js for similar changes here.

huss commented:
Thanks again to @truongdd03 for good work. There are a few comments to address but overall this looks good and testing found no issues. I did edit the PR description a little to clarify about merging in developement.

truongdd03 committed:
Some small changes

truongdd03 committed:
Add unitRepresentType for special units

truongdd03 committed:
Change the conversion from celsius to fahrenheit

truongdd03 commented:
Thank you so much for your review, @huss. I have addressed the issues and resolved all the comments. Please let me know if there's something else that needs to be changed. Thank you.

huss reviewed (approved):
Changes all look good. This is ready to merge.
